### PR TITLE
Add Trusted Types policy

### LINKED ISSUE #13228: website with csp header require-trusted-types-for 'script' don't work after release of chrome 83
# Bug report

## Describe the bug

script loading does not work for site with csp header require-trusted-types-for 'script' don't work after release of chrome 83.

this disables assigning of script.src to unsafe url (non trusted types)
[bug location] (https://github.com/zeit/next.js/blob/7eaf9b8bab4e93b891be89704399644eafe70e21/packages/next/client/page-loader.js#L265)

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. use chrome 83.
2. goto [here](https://arcane-refuge-12762.herokuapp.com/some)
3. Click here on the only link available on screen with text: clicking here will also not work
4. See error on console

## Expected behavior

it should work normally

## Screenshots
screenshot

## System information

- OS: [ macOS, Windows]
- Browser chrome
- Version of Next.js: 9.X
- Version of Node.js: [e.g. 10.10.0]

## Additional context
[doc](https://web.dev/trusted-types/)

### PR DESCRIPTION
**Pull request by TyMick**
Resolves #13228, at least for production builds (doesn't fix the bug in the dev server, seemingly [due to webpack](https://github.com/vercel/next.js/issues/13228#issuecomment-635112229)), though a security review is needed.

[Trusted Types](https://web.dev/trusted-types/) is a new experimental DOM API created to combat DOM XSS, supported now in Chrome 83, with a polyfill available for other browsers. @tipra34 noticed that when requiring Trusted Types for scripts via `<meta httpEquiv="Content-Security-Policy" content="require-trusted-types-for 'script';" />`, it broke links (using `next/link`) in Chrome 83: hovering over links would show the correct target URL, but clicking the link would do nothing. They ended up determining that the offending script src assignment took place in [`PageLoader.loadScript()`](https://github.com/vercel/next.js/blob/ad145d347f7fed2b8200708d1f9d2040e71da07c/packages/next/client/page-loader.js#L265).

This update adds a new custom Trusted Types policy by placing the following at the top of `page-loader.js`:

```js
let trustedTypesPolicy = undefined
if (window?.trustedTypes?.createPolicy) {
  trustedTypesPolicy = window.trustedTypes.createPolicy('next-trusted-types', {
    // Needs security review regarding DOM XSS
    createHTML(dirty) {
      return dirty.replace(/</g, '&lt;') // Escapes `<` characters to prevent the creation of new HTML elements
    },
    createScriptURL(dirty) {
      return new URL(dirty.replace(/</g, '&lt;'), document.baseURI)
    },
  })
}
```

...and changing the `script.src` assignment in `loadScript()` to

```js
script.src = trustedTypesPolicy
  ? trustedTypesPolicy.createScriptURL(url)
  : url
```

I'm not certain how secure `.replace(/</g, '&lt;')` is as a security policy, so more experienced eyes there would be appreciated. Alternatively, we could [use DOMPurify](https://web.dev/trusted-types/#use-a-library) instead of creating a custom policy, and just

```js
import DOMPurify from 'dompurify'
script.src = DOMPurify.sanitize(url, { RETURN_TRUSTED_TYPE: true })
```

...but that adds an extra dependency we aren't already using.

Also, this only fixes the bug for production builds (including when server-rendering). Using the dev server, the entire page breaks and the console points to a couple of different places where webpack is using `eval()`. Haven't done enough investigation yet to determine if that's something that we can fix on our end or that webpack needs to fix.

### CODE REVIEW

**Review by timneutkens — CHANGES_REQUESTED** [review]
This needs a review from @spanicker's team 👍

File: packages/next/client/page-loader.js
Comment (prateekbh) [inline packages/next/client/page-loader.js:12]: is this even needed for this particular case

**Comment by prateekbh** [comment]
LGTM except the nit noted

File: packages/next/client/page-loader.js
Comment (rictic) [reply packages/next/client/page-loader.js:12]: Looks like it isn't used, so the createHTML method here can be dropped.

File: packages/next/client/page-loader.js
Comment (rictic) [inline packages/next/client/page-loader.js:15]: So, the attack you're guarding against with createScriptUrl is an untrusted/malicious URL, which can execute arbitrary code (e.g. by loading a script from a hostile origin that exfiltrates user data, mines bitcoins, etc)

It looks like you construct the URLs in two places, at line 263 and line 103. If you are confident that the data used to construct those URLs can't come from untrusted sources, then you could have a createScriptURL function that just returns its argument, and call policy.createScriptUrl there (with comments about your assumptions).

Then, because policy isn't exported, you and your users can reduce the amount of code that needs to be subject to careful security review to just those where the policy is used in this file.

**Review by rictic — COMMENTED** [review]
Please forgive the drive-by review by a rando. For context, I've worked on integrating Trusted Types into the lit-html project. Hopefully this advice is helpful :)

**Comment by Siegrift** [comment]
Hi, I am an intern at Google and last year I focused on trusted types (TT) integration to client side frameworks and I also created an [integration for nextjs](https://github.com/Siegrift/nextjs-tt-integration/tree/legacy) but I didn't create a PR as the code changes were needed in React and Webpack. 

In your PR you missed one execution in [head-manager](https://github.com/vercel/next.js/blob/canary/packages/next/client/head-manager.js#L8). The code will be executed if you include `script` tag in the [nextjs/head](https://nextjs.org/docs/api-reference/next/head) or an `iframe` with `srcdoc` attribute. For nextjs to be TT compliant in development mode you need to patch [prerender-indicator.js](https://github.com/vercel/next.js/blob/canary/packages/next/client/dev/prerender-indicator.js#L93) and [dev-build-watcher.js](https://github.com/vercel/next.js/blob/canary/packages/next/client/dev/dev-build-watcher.js#L96) as well.

_(Note: The PR's and examples and PR's I provide in next sections might be a bit outdated)._

Unfortunately _(as you already know)_ for pleasant development with trusted types enabled you need to integrate with Webpack and React as well. There are open PRs for Webpack _(https://github.com/webpack/webpack/pull/9856 and https://github.com/webpack/webpack/pull/9861)_ and React accepted TT behind feature flag _(https://github.com/facebook/react/pull/16157)_. Adding proper support for trusted types would require finishing the integration in these dependencies _(merging webpack, removing the feature flag in React or using the build with TT)_. Using latest version of React _(v16.13.1)_ without TT enabled, you are not able to use `dangerouslySetInnerHTML` nor any potentially unsafe attributes/properties _(e.g `iframe srcdoc`)_ because the values will be stringified along the way to the sink reaching it as a plain string. Soon it might will be possible to use at least `dangerouslySetInnerHTML` (https://github.com/facebook/react/pull/17773).

As nextjs can render the code on server side, it is possible to introduce reflected XSS _([example](https://github.com/Siegrift/nextjs-tt-integration/blob/master/example-app/pages/dangerous.js))_. We also experimented with adding [TT for React on server](https://github.com/facebook/react/pull/16555) using [the polyfill](https://github.com/w3c/webappsec-trusted-types#nodejs), however this is not in the threat model of TT.

Last thing, if you want to use TT in dev mode as well, you can use [CSP in report only mode](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy-Report-Only). Unfortunately, that is the only way as you can't use [TT default policy](https://w3c.github.io/webappsec-trusted-types/dist/spec/#default-policy-hdr), because Webpack code is executed before user code.

If you are interested in adding full trusted types support, we are glad to help you with integration or reviews.

**Comment by koto** [comment]
FWIW since next.js is now in Typescript, we can leverage [tsec](https://github.com/google/tsec) to statically discover Trusted Types violations.

Where's what it shows now (there's at least one false positive violation with `setTimeout`:

```
$ cd pacakges/next
$ yarn add tsec --dev
$ yarn run tsec -p tsconfig.json
yarn run v1.22.10
$ /Users/koto/dev/next.js/node_modules/.bin/tsec -p tsconfig.json
client/route-loader.ts:147:5 - error TS21228: [ban-script-src-assignments] Do not assign variables to HTMLScriptElement#src, as this can lead to XSS.

147     script.src = src
        ~~~~~~~~~~~~~~~~
lib/recursive-delete.ts:5:25 - error TS21228: [ban-window-stringfunctiondef] Do not use setTimeout, as calling it with a string argument can cause code-injection security vulnerabilities.

5 const sleep = promisify(setTimeout)
                          ~~~~~~~~~~
client/head-manager.ts:25:7 - error TS21228: [ban-element-setattribute] Do not use Element#setAttribute or similar APIs, as this can lead to XSS.

25       el.setAttribute(attr, props[p])
         ~~~~~~~~~~~~~~~
client/head-manager.ts:31:5 - error TS21228: [ban-element-innerhtml-assignments] Assigning directly to Element#innerHTML can result in XSS vulnerabilities.

31     el.innerHTML = dangerouslySetInnerHTML.__html || ''
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/script.tsx:79:5 - error TS21228: [ban-element-innerhtml-assignments] Assigning directly to Element#innerHTML can result in XSS vulnerabilities.

79     el.innerHTML = dangerouslySetInnerHTML.__html || ''
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
client/script.tsx:81:5 - error TS21228: [ban-script-content-assignments] Do not assign values to HTMLScriptElement#text or HTMLScriptElement#textContent, as this can lead to XSS.

 81     el.textContent =
        ~~~~~~~~~~~~~~~~
 82       typeof children === 'string'
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
 85         ? children.join('')
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~
 86         : ''
    ~~~~~~~~~~~~
client/script.tsx:88:5 - error TS21228: [ban-script-src-assignments] Do not assign variables to HTMLScriptElement#src, as this can lead to XSS.

88     el.src = src
       ~~~~~~~~~~~~
client/script.tsx:97:5 - error TS21228: [ban-element-setattribute] Do not use Element#setAttribute or similar APIs, as this can lead to XSS.

97     el.setAttribute(attr, value)
       ~~~~~~~~~~~~~~~
```

React by now should support TT in `dangerouslySetInnerHtml`, Webpack 5 is also OK via [output.trustedTypes](https://webpack.js.org/configuration/output/#outputtrustedtypes), so the integration should be relatively simple.

**Comment by timneutkens** [comment]
👋 Someone from the Chrome team is going to start working on adding Trusted Types to Next.js early January. I'm going to close this PR as we can't land it as-is. I've opened a new issue for implementing trusted types here to track it: #32209

**Pull request closed without merging**
